File: CP/CP Java/src/Vidit_and_simple_queries.java
import java.util.*;

class Main1 {
    static int blk = 550;
    static int[] m = new int[200005];
    static int odd1, oddg1, even2, eveng2, cnt;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int cases = sc.nextInt();
        while (cases-- > 0) {
            int n = sc.nextInt();
            int q = sc.nextInt();
            int[] a = new int[n];
            ArrayList<Integer> temp = new ArrayList<>();

            for (int i = 0; i < n; i++) {
                a[i] = sc.nextInt();
                temp.add(a[i]);
            }
            //  Collections.sort(temp);
            Set<Integer> set = new HashSet<>(temp);
            temp.clear();
            temp.addAll(set);
            Collections.sort(temp);
            for (int i = 0; i < temp.size(); i++)
                m[i] = 0;
            for (int i = 0; i < n; i++)
                a[i] = Collections.binarySearch(temp, a[i]);

            Node qry[] = new Node[q];
            for (int i = 0; i < q; i++) {
                qry[i] = new Node();
                qry[i].l = sc.nextInt();
                qry[i].r = sc.nextInt();
                qry[i].k = sc.nextInt();
                qry[i].l--;
                qry[i].r--;
                qry[i].ind = i;
            }
            Arrays.sort(qry, new Comparator<Node>() {
                public int compare(Node a, Node b) {
                    //   if (a.l / blk != b.l / blk)
                    //     return (int)(a.l / blk - b.l / blk);
                    //   return (int)(a.r - b.r);
                    if (a.l / blk != b.l / blk)
                        return (int)(b.l / blk - a.l / blk);
                    return (int)(b.r - a.r);
                }
            });
            int lp = 0;
            int rp = -1;
            cnt = odd1 = oddg1 = even2 = eveng2 = 0;
            int[] ans = new int[q];
            for (int i = 0; i < q; i++) {
                int l = qry[i].l;
                int r = qry[i].r;
                while (lp < l) {
                    remove(a[lp]);
                    lp++;
                }
                while (l < lp) {
                    add(a[lp - 1]);
                    lp--;
                }
                while (rp < r) {
                    rp++;
                    add(a[rp]);
                }
                while (rp > r) {
                    remove(a[rp]);
                    rp--;
                }
                ans[qry[i].ind] = get(qry[i].k, r - l + 1);
            }
            for (int i = 0; i < q; i++)
                System.out.println(ans[i]);
        }
    }

    public static void add(int x) {
        if(m[x]<0){
            m[x]++;
            return;
        }
        if (m[x] == 0)
            odd1++;
        else if (m[x] == 1) {
            odd1--;
            even2++;
        } else if (m[x] == 2) {
            even2--;
            oddg1++;
        } else if (m[x] % 2 == 1) {
            oddg1--;
            eveng2++;
        } else {
            eveng2--;
            oddg1++;
        }
        m[x]++;
        if (m[x] % 2 == 1)
            cnt++;
        else
            cnt--;
    }

    static void remove(int x) {
        if(m[x]<=0){
            m[x]--;
            return;
        }
        if (m[x] == 1)
            odd1--;
        else if (m[x] == 2) {
            even2--;
            odd1++;
        } else if (m[x] == 3) {
            oddg1--;
            even2++;
        } else if (m[x] % 2 == 1) {
            oddg1--;
            eveng2++;
        } else {
            eveng2--;
            oddg1++;
        }
        m[x]--;
        if (m[x] % 2 == 1)
            cnt++;
        else
            cnt--;
    }

    static int get(int k,int len) {
        if ((Math.abs(cnt - k) & 1) != 0 || k > len)
            return -1;
        if (k == cnt)
            return 0;
        else if (k < cnt)
            return (cnt - k) / 2;
        else {
            int delta = k - cnt;
            if (delta <= 2 * (even2 + eveng2))
                return delta / 2;
            else
                return delta - eveng2 - even2;
        }
    }
}

class Node {
    int blk = 550;
    int l, r, k;
    int ind;
}

File: CP/CP Python/Vidit_and_simple_queries.py
from bisect import bisect_left
from functools import cmp_to_key
class Node:
    blk=None
    l=None
    r=None
    k=None
    ind=None
    def __init__(self) -> None:
        self.blk=550
        self.l=0
        self.r=0
        self.k=0
        self.ind=0

blk=550
m=[0]*200005
odd1=0
oddg1=0
even2=0
eveng2=0
cnt=0


def compare(a,b):
    if(a.l//blk != b.l//blk):
        return b.l//blk - a.l//blk
    return b.r-a.r

def add(x):
    global odd1,oddg1,even2, eveng2, cnt
    if m[x]<0:
        m[x]+=1
        return
    if m[x] == 0:
        odd1+=1
    elif m[x] == 1:
        odd1-=1
        even2+=1
    elif m[x] == 2:
        even2-=1
        oddg1+=1
    elif m[x] % 2 == 1:
        oddg1-=1
        eveng2+=1
    else:
        eveng2-=1
        oddg1+=1
    m[x]+=1
    if(m[x]%2==1):
        cnt +=1
    else:
        cnt-=1

def remove(x):
    global odd1,oddg1,even2, eveng2, cnt
    if m[x]<=0:
        m[x]-=1
        return
    if m[x] == 1:
        odd1-=1
    elif m[x] == 2:
        odd1+=1
        even2-=1
    elif m[x] == 3:
        even2+=1
        oddg1-=1
    elif m[x] % 2 == 1:
        oddg1-=1
        eveng2+=1
    else:
        eveng2-=1
        oddg1+=1
    m[x]-=1
    if(m[x]%2==1):
        cnt +=1
    else:
        cnt-=1

def get(k,len):
    global odd1,oddg1,even2, eveng2, cnt
    if (abs(cnt-k) & 1) != 0 or k > len:
        return -1
    if k == cnt:
        return 0
    elif k<cnt:
        return (cnt-k)//2
    else:
        delta = k-cnt
        if delta <= 2 * (even2+eveng2):
            return delta//2
        else:
            return delta - eveng2 - even2
    

for _ in range(int(input())):
    [n,q] = [int(x) for x in input().split()]
    a = [int(x) for x in input().split()]
    temp = [x for x in a]
    
    set_ = set(temp)
    temp.clear()
    temp = [x for x in set_]
    temp.sort()
    for i in range(len(temp)):
        m[i] = 0
    for i in range(n):
        a[i] = bisect_left(temp, a[i])
    qry = [0]*q
    for i in range(q):
        qry[i] = Node()
        [qry[i].l, qry[i].r, qry[i].k] = [int(x) for x in input().split()]
        qry[i].l -= 1
        qry[i].r -= 1
        qry[i].ind = i
    # sorted(qry, key=cmp_to_key(compare)) --> slow
    qry.sort(key=cmp_to_key(compare)) # --> fast
    lp=0
    rp=-1
    cnt = odd1 = oddg1 = even2 = eveng2 = 0
    ans = [0]*q
    for i in range(q):
        l = qry[i].l
        r = qry[i].r
        while lp < l:
            remove(a[lp])
            lp+=1
        while l<lp:
            add(a[lp-1])
            lp-=1
        while (rp < r):
            rp+=1
            add(a[rp])
        while (rp > r):
            remove(a[rp])
            rp-=1
        ans[qry[i].ind] = get(qry[i].k,r-l+1)
    for i in range(q):
        print(ans[i])
    
        
    
    